gpu-miner.c

```diff
@@ -148,91 +131,86 @@ int main() {
 
 	// Set OpenCL kernel arguments
 	ret = clSetKernelArg(kernel, 0, sizeof(cl_mem), (void *)&blockHeadermobj);
-	if (ret != CL_SUCCESS) { printf("failed to set first kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set first kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 1, sizeof(cl_mem), (void *)&headerHashmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set fifth kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set fifth kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 2, sizeof(cl_mem), (void *)&targmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set third kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set third kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 3, sizeof(cl_mem), (void *)&nonceOutmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set second kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set second kernel arg: \n"); exit(1); }
 	ret = clSetKernelArg(kernel, 4, sizeof(cl_mem), (void *)&nonceOutLockmobj);
-	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); return -1; }
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+	ret = clSetKernelArg(kernel, 5, sizeof(cl_mem), (void *)&iter_per_threadmobj);
+	if (ret != CL_SUCCESS) { printf("failed to set fourth kernel arg: \n"); exit(1); }
+
+	// Rough scan for 'optimal' thread count
+	double hash_rate, prev_hash_rate = 0;
+	global_item_size = 16;
+	iter_per_thread = 16 * 256;
+	while(1) {
+		global_item_size *= 2;
+		double temp = grindNonces(global_item_size, iter_per_thread);
+		while (temp == -1) {
+			// Repeat until no block is found
+			temp = grindNonces(global_item_size, iter_per_thread);
+			printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+			fflush(stdout);
+		}
+		hash_rate = temp;
+		printf("\rMining at %.3f MH/s\t%u blocks mined", hash_rate, blocks_mined);
+		fflush(stdout);
+		if (hash_rate < prev_hash_rate) {
+			break;
+		}
+		prev_hash_rate = hash_rate;
+	}
+	printf("\rRough search found %zd threads to be the best at %.3f MH/s\n", global_item_size/2, prev_hash_rate);
+	fflush(stdout);
+
+	// Now we know the optimal is betweem global_item_size and global_item_size / 2
+	// Scan intermediate 16 values and pick the highest
+	int dec = (global_item_size - global_item_size / 2) / 16;
```

Comment:
I don't consider 'dec' to be an immediately obvious variable name, I would prefer step_size or something similar